### ISSUE TITLE
WebSocketDisconnect not caught while waiting on async tasks

### ISSUE THREAD

**Issue by eboddington** [tiangolo/fastapi#3934]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.websocket("/ws")
async def my_websocket(websocket: WebSocket):
    await websocket.accept()

    try:
        await websocket.send_text("Hello!")
        await asyncio.sleep(5)
        await websocket.send_text("Finished")
    except WebSocketDisconnect:
        print("Client disconnected")
```

### Description

Example first, background below.

### Steps to reproduce using provided example
* Connect to the `/ws` WebSocket endpoint.
* Confirm client receives "Hello!" message from the server.
* Disconnect the client within 5 seconds.

### Expected result
The `WebSocketDisconnect` error is caught as soon as the client closes the connection and "Client disconnected" is printed in the console.

### Actual result
No `WebSocketDisconnect` exception is raised and instead, `websockets.exceptions.ConnectionClosedError` is raised, only **when the server attempts to send a message back to the client.**

With debug logging enabled, I can see the following **at the exact moment the client disconnects**:
```
DEBUG:websockets.protocol:server - event = data_received(<6 bytes>)
DEBUG:websockets.protocol:server < Frame(fin=True, opcode=<Opcode.CLOSE: 8>, data=b'', rsv1=False, rsv2=False, rsv3=False)
DEBUG:websockets.protocol:server - state = CLOSING
DEBUG:websockets.protocol:server > Frame(fin=True, opcode=<Opcode.CLOSE: 8>, data=b'', rsv1=False, rsv2=False, rsv3=False)
DEBUG:websockets.protocol:server x half-closing TCP connection
DEBUG:websockets.protocol:server - event = eof_received()
DEBUG:websockets.protocol:server - event = connection_lost(None)
DEBUG:websockets.protocol:server - state = CLOSED
DEBUG:websockets.protocol:server x code = 1005, reason = [no reason]
```

So, the `websockets` library does receive an event as soon as the connection is closed. But this is not propagated to a `WebSocketDisconnect` error and I've found no way to catch this as it happens.

## Background
In my real code, at some point, the server must `await` a long-running task and send a message to the client when it has finished. This task is potentially expensive - as such, it is strongly preferable to cancel it as soon as a client disconnects, instead of waiting for it to finish only to find out at that point that the client is no longer there.

In my testing, I've seen that the `WebSocketDisconnect` error **is** raised in some cases, of course, for example:
```
@app.websocket("/ws")
async def my_websocket(websocket: WebSocket):
    await websocket.accept()

    try:
        await websocket.receive_text()
    except WebSocketDisconnect:
        print("Client disconnected")
```

If the client connects and then disconnects before sending text, then the disconnection is caught successfully. But, not if the client disconnects during a task that the server is waiting on.

### Operating System

macOS

### FastAPI Version

0.68.0

### Python Version

3.9.5

**Comment by dstlny** [comment]
I think you might be able to use `asyncio.create_task` or one of the other convenient asyncio builtins, to do this?

Something like:
```python

@app.websocket("/ws")
async def my_websocket(websocket: WebSocket):
    await websocket.accept()
    task = None

    try:
        text = websocket.receive_text()
  
        task = asyncio.create_task(some_long_running_task, **kwargs)
        
    except WebSocketDisconnect:

        if task:
            task.cancel()
        print("Client disconnected")
```

**Comment by eboddington** [comment]
> I think you might be able to use `asyncio.create_task` (https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task)

Yes, ideally I would use an approach similar to this in order to cancel the task if a client disconnects midway through. But the issue is that the exception is not raised if the client disconnects at that point, meaning code like your example wouldn't work.

I've been unable to find any method of detecting the disconnect until after the task finishes (no difference when using `create_task` unfortunately), at which point cancelling becomes irrelevant. So your example is exactly what I would like to achieve, it just doesn't work.

**Comment by trondhindenes** [comment]
I just had the same issue. What you can do, is to create two tasks: one for your long-running thing and one for awaiting client text. The latter will raise an exception the moment the client disconnects, so you can use `asyncio.FIRST_COMPLETED` to return out of the task awaiter. Here's how it could look (my long-running task is a blocking redis call that waits for a list):

```python
async def get_receive_text(websocket) -> Tuple[Union[str, None], bool]:
    """The second return value bolean will return True if the client disconnected"""
    try:
        return await websocket.receive_text(), False
    except WebSocketDisconnect as e:
        return None, True


async def get_redis_data(list_name) -> Tuple[Union[str, None], bool]:
    """The second return value bolean will never return true, just there for task consistency"""
    redis = aioredis.from_url("redis://localhost:6379")
    async with redis.client() as conn:
        result = await conn.blpop(keys=[list_name], timeout=30)
        if result:
            return result[1].decode(), False
        return None, False


@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    conn_id = random.randint(0, 999)
    print(f'{conn_id=}')
    client_is_connected = True
    while client_is_connected:
        # Wait for either text from the websocket channel, OR the redis list (which will ultimately time out)
        task1 = get_receive_text(websocket)
        task2 = get_redis_data(list_name=conn_id)
        if websocket.client_state.name != 'CONNECTED':
            client_is_connected = False
            break
        # asyncio wait will return when either text is received from the client, 
        # or the redis blocking call returned data
        # or the redis blocking call timed out
        # or the client disconnected
        done, pending = await asyncio.wait([task1, task2], return_when=asyncio.FIRST_COMPLETED)
        done_task = done.pop()
        while len(pending) > 0:
            # cancel the other tasks in the waiter
            pending_task = pending.pop()
            pending_task.cancel()
        result, connection_closed = done_task.result()
        if connection_closed:
            client_is_connected = False
            break
        if result:
            print(result)
            await websocket.send_text(result)

```

**Comment by goDeni** [comment]
There is no correct solution if your coroutine only send data and not receive 

Only trying to receive data to sync websocket state?..


But this is not good solution because why more simpler raise cancell task exception for this coroutine
And debug logs know about connection lost


Is there a plan to fix this issue in the future?
@tiangolo 

Thanks for answer in advance.

**Comment by danieldaeschle** [comment]
Having the same issue

**Comment by linpan** [comment]
Having the same issue, tooo

**Comment by joshjhans** [comment]
This works well! One concern I am confused about is how to ensure all clients subscribed to the same websocket receive all messages; currently only alternating clients receives a new message

**Comment by trondhindenes** [comment]
> This works well! One concern I am confused about is how to ensure all clients subscribed to the same websocket receive all messages; currently only alternating clients receives a new message

you need to track subscribed clients in a db such as redis etc, and have each message "broadcast" to each subscriber (aka a for each loop basically)

**Comment by joshjhans** [comment]
I cannot provide many details, but this implementation misses sending messages when get_recieve_data adds a message to the Redis list being read by get_redis_data. Still investigating...

**Comment by JarroVGIT** [comment]
@joshjhans that sounds like another issue than what this issue is relating to. If you have more information, I would recommend opening another issue (if it is FastAPI related).
